docs/architecture/psa-migration/psa-legacy-bridges.md

```diff
@@ -0,0 +1,321 @@
+
+There are several scenarios where an application has a legacy key pair or public key (`mbedtls_pk_context`) and needs to create a PSA key object (`psa_key_id_t`).
+
+Reasons for creating a legacy key object, where it's impossible or impractical to directly create a PSA key:
+
+* A very common case where the input is a legacy key object is parsing. PSA does not (yet) have an equivalent of the `mbedtls_pk_parse_xxx` functions.
+* The PSA key creation interface is less flexible in some cases. In particular, PSA RSA key generation does not (yet) allow choosing the public exponent.
+* The pk object may be created by a part of the application (or a third-party library) that hasn't been migrated to the PSA API yet.
+
+Reasons for needing a PSA key object:
+
+* Using the key in TLS 1.3 or some third-party interface that takes a PSA key identifier as input.
+* Benefiting from a PSA accelerator, or from PSA's world separation, even without `MBEDTLS_USE_PSA_CRYPTO`. (Not a priority scenario: we generally expect people to activate `MBEDTLS_USE_PSA_CRYPTO` at an early stage of their migration to PSA.)
+
+Gap: a way to create a PSA key object from an `mbedtls_pk_context`. This partially exists in the form of `mbedtls_pk_wrap_as_opaque`, but it is not fully satisfactory, for reasons that are detailed in “[API to create a PSA key from a PK context](#api-to-create-a-psa-key-from-a-pk-context)” below.
+
+#### Using a PSA key as a PK context
+
+There are several scenarios where an application has a PSA key and needs to use it through an interface that wants an `mbedtls_pk_context` object. Typically, there is an existing key in the PSA key store (possibly in a secure element and non-exportable), and the key needs to be used in an interface that requires a `mbedtls_pk_context *` input, such as Mbed TLS's X.509 API or a similar third-party interface, or the `mbedtls_pk_write_xxx` interfaces which do not (yet) have PSA equivalents.
+
+There is a function `mbedtls_pk_setup_opaque` that mostly does this. However, it has several limitations:
+
+* It creates a PK key of type `MBEDTLS_PK_OPAQUE` that wraps the PSA key. This is good enough in some scenarios, but not others. For example, it's ok for pkwrite, because we've upgraded the pkwrite code to handle `MBEDTLS_PK_OPAQUE`. That doesn't help users of third-party libraries that haven't yet been upgraded.
+* It ties the lifetime of the PK object to the PSA key, which is error-prone: if the PSA key is destroyed but the PK object isn't, there is no way to reliably detect any subsequent misuse of the PK object.
+* It is only available under `MBEDTLS_USE_PSA_CRYPTO`. (Not a priority concern: we generally expect people to activate `MBEDTLS_USE_PSA_CRYPTO` at an early stage of their migration to PSA.)
+
+Gap: a way to copy a PSA key into a PK context. This can only be expected to work if the PSA key is exportable.
+
+[OPEN] Is `mbedtls_pk_setup_opaque` ok or do we want to tweak it?
```

Comment:
I'm not convinced we can tweak it to avoid the lifetime restriction: if the policy of the original PSA keys allows neither exporting not copying, then we have no way of making the PK object independent. I'll note that's not an edge case, but actually the primary use case (hint: the name) - you have a key in a secure element and want to use it with TLS or X.509, which want a PK context. So you want a PK context that's really just a wrapper around your PSA key.

Another approach would be to try detecting misuse but I'm not sure what we can to here:
- Simple case: the PSA key is `destroy()`ed and we try using the PK context - I think this should result in clean errors, though I don't think we're currently testing this and we probably should.
- Hard case: the PSA key is `destroy()`ed then another key is created which ends up having the same ID, then we try using the PK context - not sure what we can do about that without special support from the PSA implementation (we would need the `destroy()` call to somehow notify the PK context, a bit like what happens (or doesn't happen yet?) for multi-part operations).

I'll also note that IMO the lifetime restriction become less of a problem with the introduction of `pk_copy_from_psa()`, as the two complement each other:
- With `pk_copy_from_psa()` you get a normal PK object that stands on its own, has a normal PK type, etc. (IMO that includes being unrestricted by the policies from the original key.)
- With `pk_setup_opaque()` you get a special PK object that's actually just a wrapper for PSA key: it has a special type, does not stand on its own (lifetime restriction), obeys the policies of the PSA key...

Reply:
I'm not sure what the point of your comment is. Are you arguing that we need two different functions, one that wraps and one that copies? That's exactly what I'm proposing. Maybe it wasn't clear at this point in the document? If so, sorry about the confusion. If not, what is your objection here?

Reply:
Sorry, the point of my comment was that about the question "Is `mbedtls_pk_setup_opaque` ok or do we want to tweak it?" I think yes it's OK and no we don't want to tweak it (and fwiw I'm not convinced we could easily even if we wanted to).

> Are you arguing that we need two different functions, one that wraps and one that copies? That's exactly what I'm proposing.

I realise that's what you're proposing and I fully agree! I was just saying that's another reason why I think `pk_setup_opaque()` is fine as it is.